### PR TITLE
Add a provider wrapper to multiplex callbacks

### PR DESCRIPTION
**Pull request by alsamylkin**
Some providers(like MockDynamicProvider) allow to register only one callback. We can easily extend them by wrapping in another provider, that maintains a slice of callbacks that would to be executed on a change.

### CODE REVIEW

File: config/stack_callback_provider.go [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+// THE SOFTWARE.
+
+package config
+
+import (
+	"fmt"
+	"sync"
+)
+
+type stackCallbackProvider struct {
```
Comment (alsamylkin) [inline config/stack_callback_provider.go]: I am open to a better name :)

Comment (dmcaulay) [inline config/stack_callback_provider.go]: `mutliCallbackProvider`?

**Comment by dmcaulay** [comment]
this looks good to me!

**Comment by glibsm** [comment]
I think it may just be the late hour, but the reasoning for the particular approach is eluding me

I think it would be better to provide a way for dynamic providers to embed a struct that manages these callbacks. 

Wrapping it into a private struct is not ideal, as at that point control over the provider is lost somewhat - we just returned a private instance.

**Comment by alsamylkin** [comment]
@glibsm I don't see, how do we loose a control over a provider, we still have the instance that we are wrapping, so another provider can embed this one and use it together with a provider that doesn't allow multiple callbacks per key.

I do think though, that we shouldn't make a user to register callbacks manually, Value can do it inside, but there are lots of things to consider: we copy Values a lot, so registering callbacks there can potentially explode our memory usage, concurrent safety is not here yet, etc.

File: config/stack_callback_provider.go (hunk 1 above)
Comment (glibsm) [inline config/stack_callback_provider.go]: +1 for multi

**Comment by glibsm** [comment]
@alsamylkin my suggestion was to consider using composition instead. I should be easy for a dynamic provider to allow this out of the box, without the user having to wrap their providers.

I don't think it's an easy change, and certainly not API compatible.

**Pull request merged**